### ISSUE TITLE
port to Python 3

### ISSUE THREAD

**Issue by pabs3** [mnauw/git-remote-hg#34]
Now that mercurial has been ported to Python 3 it would be useful to have git-remote-hg work with Python 3, especially since distributions like Debian are removing Python 2. Currently the version of git-remote-hg in Debian just crashes since mercurial in Debian no longer provides Python 2 support.

https://bugs.debian.org/943032

**Comment by mnauw** [comment]
The port to python3 has already been done in latest master/version, see also issue #27.  However, Debian has not yet moved to that version, which would be the way to go.

**Comment by pabs3** [comment]
Hmm, I still see the scripts using Python 2 instead of Python 3:

git-remote-hg (master=) $ head -n1 git-remote-hg git-hg-helper
==> git-remote-hg <==
#!/usr/bin/env python

==> git-hg-helper <==
#!/usr/bin/env python

-- 
bye,
pabs

https://wiki.debian.org/PaulWise

**Comment by mnauw** [comment]
Note quite, that python shebang situation is a bit of a mess for scripts or distributions (see also e.g. https://www.python.org/dev/peps/pep-0394).  In practice, note that python is nowadays often a symlink to python2(.x), but it may well be(come) one to python3 (depending on whether distro considers that the main/primary python; see mentioned PEP).

Like Mercurial, `git-remote-hg` has not just been "ported", but has been made compatible and dual-executable (so to speak).  And so that script shebang is meant as a "universal" shebang that indicates it is compatible with both python2 and python3 (whatever it happens to be symlinked to).  Of course, practices, customs and opinions on the "correct" way to deal with shebang/compatibility may considerably vary (that is actually the mess).  In any case, the same shebang is used in Mercurial's `hg` script.   So however distro packaging handles that, so it should do so in this case (or any python2/3 case).

Simply put, it seems up to the distro packaging to straighten out the shebang according to how python2/3 is addressed.  This is also what `pip` or `pip3` will do when installed using those (in a virtualenv or not); the shebang will be adjusted accordingly (however unfortunate such "tricks" might be).

**Comment by pabs3** [comment]
Thanks for the link to PEP 394, sigh. I'll forward this to the
maintainer of the Debian git-remote-hg package.

-- 
bye,
pabs

https://bonedaddy.net/pabs3/
